include/bit_algorithm_details.hpp

```diff
@@ -132,6 +132,119 @@ ForwardIt word_shift_right(ForwardIt first,
+bool _is_one_past_last_bit(bit_iterator<It> maybe_end, It&& base) {
+    return maybe_end.position() == 0 
+      && std::next(std::forward<It>(base)) == maybe_end.base();
+}
+
+// checks that two bit iterators point to the same word
+template <class It>
+constexpr bool _in_same_word(bit_iterator<It> lhs, bit_iterator<It> rhs) {
+    return lhs.base() == rhs.base();
+}
+
+// simple alias for right shift
+template <class WordType>
+WordType _shift_towards_lsb(WordType word, std::size_t n) {
+    return word >> n; 
+} 
+
+// simple alias for left shift
+template <class WordType>
+WordType _shift_towards_msb(WordType word, std::size_t n) {
+    return word << n;
+}
+
+/* Used to read partial/full words and pad any missing digits. Will not
+ * read outside of the word pointed to by the first iterator (see case 4) 
+ *
+ * Case 0: 01011101
+ *        L       F
+ * Case 1: 01011101 -> padded with 0s -> 00001101
+ *            L   F
+ * Case 2: 01011101 -> padded with 1s -> 01011111 
+ *        L    F 
+ * Case 3: 01011101 -> padded with 0s -> 00011100
+ *           L  F
+ * Case 4: 01011101 11111111 -> treated as 01011101
+ *           F           L                L  F
+ *
+ * Note: word is read from [first, last), meaning the element pointed
+ * to by last is not included in the read. if first == last, behavior
+ * is undefined
+ */
+template <class It>
+typename bit_iterator<It>::word_type _padded_read(bit_iterator<It> first, 
+    bit_iterator<It> last, const bit::bit_value bv) {
+
+    using word_type = typename bit_iterator<It>::word_type;
+
+    constexpr std::size_t num_digits = binary_digits<word_type>::value;
+    const std::size_t first_position = first.position();
+    const std::size_t last_position = last.position();
+    const word_type read = *(first.base());
+    constexpr word_type all_ones = _all_ones();
+
+    word_type mask;
+
+    if (_is_aligned_lsb(first)) {
+        if (_in_same_word(first, last)) {
+            // Case 1
+            if (bv == bit0) {
+                mask = _shift_towards_lsb(all_ones, num_digits - last_position);
+                return read & mask;
+            } else {
+                mask = _shift_towards_msb(all_ones, last_position); 
+                return read | mask;
+            }
+        } else {
+            // Case 0
+            return read;
+        }
+    } else {
+        if (!_in_same_word(first, last)) {
+            // Case 2
+            if (bv == bit0) {
+                mask = _shift_towards_msb(all_ones, first_position); 
+                return read & mask;
+            } else {
+                mask = _shift_towards_lsb(all_ones, num_digits - first_position); 
+                return read | mask;
+            }
+        } else {
+            // Case 3
+            if (bv == bit0) {
+                mask = _shift_towards_msb(all_ones, first_position); 
+                mask &= _shift_towards_lsb(all_ones, num_digits - last_position);
+                return read & mask;
```

Comment:
Same idea here. It ends up only saving us a single instruction probably. 

Essentially we no longer need the `& mask` but at the cost of losing the original position of the bits.

Reply:
And for the version that does end up masking, the canonical way to do this is 

`mask = ((1<<first_pos)-1) << (last_pos-first_pos);` 
as opposed to 
`mask = ((-1) << first_pos) & ((-1) >> (digits-last_pos));`

which [saves a whopping **TWO** instructions!  ](https://godbolt.org/z/1xdQ4q)

Reply:
[I just tested all 3 out.](https://godbolt.org/z/69LOSr) Shifting input word to shave off the edges uses the least instructions.